bin/cron.js

```diff
@@ -38,6 +38,252 @@ cron.schedule("0 0 0 * * *", () => {
+ */
+cron.schedule("0 0 * * Mon", async () => {
+  const logs = {};
+  Object.keys(config.mongoDBs).map(async (courseID) => {
+    console.log(`Updating ${config.mongoDBs[courseID]}'s user progress`);
+    try {
+      const assignmentIdToType = {}, // Maps a course's modules assignment id to its type - e.g 22657: "practice"
+        badgeIdToPoints = {}; // Maps a course's badges id to its points - e.g 1: 200
+
+      const db = mongo.client.db(config.mongoDBs[courseID]),
+        userSubmissionsPromise = () =>
+          canvas.getSubmissions(
+            courseID,
+            "student_ids[]=all&workflow_state=graded&grouped=true&per_page=1000" // Get submissions grouped by user - e.g [{user_id:1, submissions: []}, ...]
+          ),
+        modulesPromise = () => db.collection("modules").find().sort({ _id: 1 }).toArray(),
+        badgesPromise = () => db.collection("badges").find().sort({ _id: 1 }).toArray();
+
+      // Retrieve all necessary information at once
+      const [userSubmissions, modules, badges] = await Promise.allSettled([
+        userSubmissionsPromise(),
+        modulesPromise(),
+        badgesPromise(),
+      ]);
+
+      //Intitalizes dicts of scores for practice and quizzes
+      let highestPractice = {};
+      let highestQuiz = {};
+      // Keeps track of earliest submitted time for each module in order(practice, apply, reflection)
+      let times = [{}, {}, {}];
+
+      // Initialize maps to each value
+      modules.value.map((module) => {
+        assignmentIdToType[module.practice_link] = {
+          type: "practice",
+          moduleID: module._id,
+        };
+        assignmentIdToType[module.quiz_link] = {
+          type: "apply",
+          moduleID: module._id,
+        };
+        assignmentIdToType[module.reflection_link] = { type: "reflection", moduleID: module._id };
+        //initializes dicts with module ids
+        highestPractice[module._id] = [];
+        highestQuiz[module._id] = [];
+        times[0][module._id] = { user: {}, time: 0 };
+        times[1][module._id] = { user: {}, time: 0 };
+        times[2][module._id] = { user: {}, time: 0 };
+      });
+      badges.value.map((badge) => {
+        badgeIdToPoints[badge._id] = parseInt(badge.Points);
+      });
+
+      //stores user's module progress for later
+      moduleScores = {};
+      for (const user of userSubmissions.value) {
+        if (user.submissions.length <= 0) continue;
+
+        let score = 0; // User score
+        // Number of completed assignments
+        const completed = {
+            practice: 0,
+            apply: 0,
+            reflection: 0,
+            daily: 0,
+            inspirer: 0,
+          },
+          userProgress = await db // Get current user's progress from MongoDB
+            .collection("user_progress")
+            .findOne({ user: user.user_id.toString() });
+        if (!userProgress) continue;
+        score += await updateModuleProgress(
```

Comment:
Why `updateModuleProgress` here?

Reply:
I was having trouble figuring out how to set the score properly with how it is set up, came up with the easiest way is to set up a dict that maps the score accumulated from each user's module progress, and then add the badge scores below to that score. Lmk if you can think of an easier way though